liblwgeom/lwgeom.c

```diff
@@ -2395,3 +2396,52 @@ lwgeom_is_trajectory(const LWGEOM *geom)
 	return lwline_is_trajectory((LWLINE*)geom);
 }
 
+static uint8_t
+bits_for_precision(uint8_t significant_digits)
+{
+	if (significant_digits < 1)
+		lwerror("Must have at least one significant digit");
+
+	if (significant_digits > 15)
+		lwerror("Can't request more than 15 significant digits");
+
+	return (uint8_t) ceil(significant_digits / log10(2));
+}
+
+static inline
+double mask_double(double d, int64_t mask)
+{
+	int64_t* double_bits = (int64_t*) (&d);
+
+	(*double_bits) &= mask;
+
+	return *((double*) double_bits);
+}
+
+void lwgeom_trim_bits_in_place(LWGEOM* geom, uint8_t sig_x, uint8_t sig_y, uint8_t sig_z, uint8_t sig_m)
+{
+	LWPOINTITERATOR* it = lwpointiterator_create_rw(geom);
+	POINT4D p;
+	uint8_t bits_x = bits_for_precision(sig_x);
+	uint8_t bits_y = bits_for_precision(sig_y);
+	uint8_t bits_z = lwgeom_has_z(geom) ? bits_for_precision(sig_z) : 52;
+	uint8_t bits_m = lwgeom_has_m(geom) ? bits_for_precision(sig_m) : 52;
+	int64_t mask_x = 0xffffffffffffffff << (52 - bits_x);
```

Comment:
Consider using `uint64_t` here and in the `mask_double` function to avoid sign conversions (not that important here, but it shuts up the compiler).